extensions/amp-embedly/0.1/amp-embedly.js

```diff
@@ -0,0 +1,190 @@
+import {dict} from '../../../src/utils/object';
+import {getIframe} from '../../../src/3p-frame';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {user} from '../../../src/log';
+
+/** @private @const {string} */
+const TAG = 'amp-embedly';
+
+/**
+ * @const {string}
+ */
+const BASE_API_URL = 'https://api.embedly.com/1/oembed?';
+
+/**
+ * Regex used to extract src from returned oEmbed data.
+ * @const {RegExp}
+ * */
+const SRC_REGEXP = /src="([^"]+)"/;
+
+/**
+ * oEmbed resource types.
+ * @const {Object<{PHOTO: string, VIDEO: string, LINK: string, RICH: string}>}
+ */
+const resourceType = Object.freeze({
+  PHOTO: 'photo',
+  VIDEO: 'video',
+  LINK: 'link',
+  RICH: 'rich',
+});
+
+/**
+ * Implementation of the amp-embedly component.
+ * See {@link ../amp-embedly.md} for the spec.
+ */
+export class AmpEmbedly extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @const @private {!../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = Services.xhrFor(this.win);
+
+    /** @private {?string}  */
+    this.key_ = null;
+
+    /** @private {?string}  */
+    this.url_ = null;
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.key_ = user().assert(
+        this.element.getAttribute('data-key'),
+        `The data-key attribute is required for <${TAG}> %s`,
+        this.element
+    );
+
+    this.url_ = user().assert(
+        this.element.getAttribute('data-url'),
+        `The data-url attribute is required for <${TAG}> %s`,
+        this.element
+    );
+
+    return this.getOembedData_()
+        .then(this.getIframe_.bind(this))
+        .then(iframe => {
+          this.iframe_ = iframe;
+          this.element.appendChild(iframe);
+
+          return this.loadPromise(iframe);
+        });
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * Fetches oEmbed data from embedly's api.
+   * @return {!Promise}
+   * */
+  getOembedData_() {
+    const params = dict({'key': this.key_, 'url': this.url_});
+    const url = addParamsToUrl(BASE_API_URL, params);
+
+    return this.xhr_.fetchJson(url, {
+      requireAmpResponseSourceOrigin: false,
+    }).then(res => res.json());
+  }
+
+  /**
+   * Gets component iframe with set source based on data type.
+   *
+   * For RICH and VIDEO types, when oEmbed response contains:
+   * - html iframe tag: we render its source as the src of our 3p iframe.
+   * - html with a script tag: we render the html and add the script to the
+   *   3p iframe document.
+   *
+   * @param {!JsonObject<{type: string, html: string, url: string}>} data
+   * @returns {Element}
+   * @private
+   */
+  getIframe_(data) {
+    const iframe = getIframe(this.win, this.element, 'embedly');
+    this.applyFillContent(iframe);
+
+    /** @type {string} */
+    let src;
```

Comment:
I strongly suggest instead hoisting the `src` resolution method into a separate method/util so that you can simply `return` the value and declare `src` as a constant.